### PR TITLE
Split V2.ShortCausalHash from V2.ShortBranchHash

### PR DESCRIPTION
**Pull request by ChrisPenner**
## Overview

Currently there's `V1.ShortBranchHash` which seems to refer exclusively to Causal Hashes, but `V2.ShortBranchHash` seems to refer to _either_ a namespace hash or causal hash depending on context.  We need to be careful to keep them separate, they're not interchangeable.

It seems all of our existing interfaces don't ever refer to namespace hashes directly, but this may change with the Namespace Diff APIs so I want to ensure no-one can mix up ShortNamespaceHash with ShortCausalHash accidentally.

## Implementation notes

* Split the V2 types up into separate types for Namespace vs Causal hashes in the v2 module, update the calling code.
* Rename `V1.ShortBranchHash -> V1.ShortCausalHash` for clarity.
* Rename some utility functions (E.g. branchHashesByPrefix)  for clarity.
* No meaningful code or behaviour changes.

I'm generally moving away from using `Branch` and instead using namespace or causal respectively, since Branch is pretty overloaded and will likely get worse if we implement "feature branches" in projects.

- [ ] After merging, pull into Enlil and do the requisite renamings there.

### CODE REVIEW

File: parser-typechecker/src/Unison/Codebase/Editor/RemoteRepo.hs [hunk 1]
```diff
@@ -110,15 +110,15 @@ data ReadRemoteNamespace
 
 data ReadGitRemoteNamespace = ReadGitRemoteNamespace
   { repo :: ReadGitRepo,
-    sbh :: Maybe ShortBranchHash,
+    sbh :: Maybe ShortCausalHash,
```
Comment (mitchellwrosen) [inline parser-typechecker/src/Unison/Codebase/Editor/RemoteRepo.hs]: sbh 🙈

File: parser-typechecker/src/Unison/Codebase/ShortCausalHash.hs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+toString = Text.unpack . toText
+
+toHash :: Coercible Hash.Hash h => ShortCausalHash -> Maybe h
+toHash = fmap coerce . Hash.fromBase32Hex . toText
+
+fromHash :: Coercible h Hash.Hash => Int -> h -> ShortCausalHash
+fromHash len =
+  ShortCausalHash . Text.take len . Hash.base32Hex . coerce
+
+-- abc -> SBH abc
```
Comment (mitchellwrosen) [inline parser-typechecker/src/Unison/Codebase/ShortCausalHash.hs]: SBH

File: parser-typechecker/src/Unison/Codebase/SqliteCodebase/Conversions.hs [hunk 3]
```diff
@@ -57,8 +57,8 @@ import qualified Unison.Util.Star3 as V1.Star3
 import qualified Unison.Var as Var
 import qualified Unison.WatchKind as V1.WK
 
-sbh1to2 :: V1.ShortBranchHash -> V2.ShortBranchHash
-sbh1to2 (V1.ShortBranchHash b32) = V2.ShortBranchHash b32
+sbh1to2 :: V1.ShortCausalHash -> V2.ShortCausalHash
```
Comment (mitchellwrosen) [inline parser-typechecker/src/Unison/Codebase/SqliteCodebase/Conversions.hs]: sbh

File: unison-cli/src/Unison/Cli/MonadUtils.hs [hunk 4]
```diff
@@ -92,8 +92,8 @@ import Unison.Codebase.Patch (Patch (..))
 import qualified Unison.Codebase.Patch as Patch
 import Unison.Codebase.Path (Path, Path' (..))
 import qualified Unison.Codebase.Path as Path
-import Unison.Codebase.ShortBranchHash (ShortBranchHash)
-import qualified Unison.Codebase.ShortBranchHash as SBH
+import Unison.Codebase.ShortCausalHash (ShortCausalHash)
+import qualified Unison.Codebase.ShortCausalHash as SBH
```
Comment (mitchellwrosen) [inline unison-cli/src/Unison/Cli/MonadUtils.hs]: SBH

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs [hunk 5]
```diff
@@ -1440,8 +1440,8 @@ loop e = do
             DebugNameDiffI fromSBH toSBH -> do
```
Comment (mitchellwrosen) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput.hs]: SBH

File: unison-cli/src/Unison/Codebase/Editor/Input.hs [hunk 6]
```diff
@@ -28,8 +28,8 @@ import Unison.Codebase.Path (Path')
 import qualified Unison.Codebase.Path as Path
 import qualified Unison.Codebase.Path.Parse as Path
 import Unison.Codebase.PushBehavior (PushBehavior)
-import Unison.Codebase.ShortBranchHash (ShortBranchHash)
-import qualified Unison.Codebase.ShortBranchHash as SBH
```
Comment (mitchellwrosen) [inline unison-cli/src/Unison/Codebase/Editor/Input.hs:32]: SBH

File: unison-cli/src/Unison/Codebase/Editor/Output.hs [hunk 7]
```diff
@@ -37,8 +37,8 @@ import Unison.Codebase.Path (Path')
 import qualified Unison.Codebase.Path as Path
 import Unison.Codebase.PushBehavior (PushBehavior)
 import qualified Unison.Codebase.Runtime as Runtime
-import Unison.Codebase.ShortBranchHash (ShortBranchHash)
-import qualified Unison.Codebase.ShortBranchHash as SBH
+import Unison.Codebase.ShortCausalHash (ShortCausalHash)
+import qualified Unison.Codebase.ShortCausalHash as SBH
```
Comment (mitchellwrosen) [inline unison-cli/src/Unison/Codebase/Editor/Output.hs]: SBH

File: unison-cli/src/Unison/Codebase/Editor/UriParser.hs [hunk 8]
```diff
@@ -336,7 +336,7 @@ parseGitProtocol =
 --
 -- >>> P.parseMaybe namespaceHashPath "."
 -- Just (Nothing,)
-namespaceHashPath :: P (Maybe ShortBranchHash, Path)
+namespaceHashPath :: P (Maybe ShortCausalHash, Path)
 namespaceHashPath = do
   sbh <- P.optional shortBranchHash
```
Comment (mitchellwrosen) [inline unison-cli/src/Unison/Codebase/Editor/UriParser.hs]: sbh

File: unison-cli/src/Unison/Codebase/Editor/UriParser.hs [hunk 9]
```diff
@@ -364,8 +364,8 @@ gitTreeishSuffix = P.label "git treeish" . P.try $ do
   void $ C.char ':'
   P.takeWhile1P (Just "not close paren") (/= ')')
 
-shortBranchHash :: P ShortBranchHash
+shortBranchHash :: P ShortCausalHash
 shortBranchHash = P.label "short branch hash" $ do
```
Comment (mitchellwrosen) [inline unison-cli/src/Unison/Codebase/Editor/UriParser.hs]: `shortBranchHash` / `"short branch hash"`

File: unison-cli/src/Unison/CommandLine/InputPatterns.hs [hunk 10]
```diff
@@ -1976,8 +1976,8 @@ debugNameDiff =
       parse =
         ( \case
             [from, to] -> first fromString $ do
-              fromSBH <- Input.parseShortBranchHash from
-              toSBH <- Input.parseShortBranchHash to
+              fromSBH <- Input.parseShortCausalHash from
```
Comment (mitchellwrosen) [inline unison-cli/src/Unison/CommandLine/InputPatterns.hs]: SBH

File: unison-cli/src/Unison/CommandLine/OutputMessages.hs [hunk 11]
```diff
@@ -73,8 +73,8 @@ import qualified Unison.Codebase.Patch as Patch
 import qualified Unison.Codebase.Path as Path
 import qualified Unison.Codebase.PushBehavior as PushBehavior
 import qualified Unison.Codebase.Runtime as Runtime
-import Unison.Codebase.ShortBranchHash (ShortBranchHash)
-import qualified Unison.Codebase.ShortBranchHash as SBH
+import Unison.Codebase.ShortCausalHash (ShortCausalHash)
+import qualified Unison.Codebase.ShortCausalHash as SBH
```
Comment (mitchellwrosen) [inline unison-cli/src/Unison/CommandLine/OutputMessages.hs]: SBH

File: unison-cli/src/Unison/CommandLine/OutputMessages.hs [hunk 12]
```diff
@@ -369,7 +369,7 @@ notifyNumbered o = case o of
         branchHashes = (fst <$> reversedHistory) <> tailHashes
      in (msg, displayBranchHash <$> branchHashes)
     where
-      toSBH :: Branch.CausalHash -> ShortBranchHash
+      toSBH :: Branch.CausalHash -> ShortCausalHash
```
Comment (mitchellwrosen) [inline unison-cli/src/Unison/CommandLine/OutputMessages.hs]: SBH

File: unison-cli/src/Unison/CommandLine/OutputMessages.hs [hunk 13]
```diff
@@ -1470,7 +1470,7 @@ notifyUser dir o = case o of
                 ""
               ]
           _ -> mempty
-      renderEntry3Column :: UTCTime -> (Maybe UTCTime, SBH.ShortBranchHash, Text) -> [Pretty]
+      renderEntry3Column :: UTCTime -> (Maybe UTCTime, SBH.ShortCausalHash, Text) -> [Pretty]
       renderEntry3Column now (mayTime, sbh, reason) =
```
Comment (mitchellwrosen) [inline unison-cli/src/Unison/CommandLine/OutputMessages.hs]: sbh

File: unison-cli/src/Unison/CommandLine/OutputMessages.hs [hunk 14]
```diff
@@ -1914,7 +1914,7 @@ prettyRelative = P.blue . P.shown
 prettyAbsolute :: Path.Absolute -> Pretty
 prettyAbsolute = P.blue . P.shown
 
-prettySBH :: IsString s => ShortBranchHash -> P.Pretty s
+prettySBH :: IsString s => ShortCausalHash -> P.Pretty s
```
Comment (mitchellwrosen) [inline unison-cli/src/Unison/CommandLine/OutputMessages.hs]: SBH

File: unison-cli/tests/Unison/Test/UriParser.hs [hunk 15]
```diff
@@ -19,13 +19,13 @@ import Unison.Codebase.Editor.RemoteRepo
...
-import Unison.Codebase.ShortBranchHash (ShortBranchHash (..))
+import Unison.Codebase.ShortCausalHash (ShortCausalHash (..))
 import Unison.NameSegment (NameSegment (..))
 
 test :: Test ()
 test = scope "uriparser" . tests $ [testShare, testGit]
 
-gitHelper :: (ReadGitRepo, Maybe ShortBranchHash, Path) -> ReadRemoteNamespace
+gitHelper :: (ReadGitRepo, Maybe ShortCausalHash, Path) -> ReadRemoteNamespace
 gitHelper (repo, sbh, path) = ReadRemoteNamespaceGit (ReadGitRemoteNamespace repo sbh path)
```
Comment (mitchellwrosen) [inline unison-cli/tests/Unison/Test/UriParser.hs]: sbh

File: unison-cli/tests/Unison/Test/UriParser.hs [hunk 16]
```diff
@@ -120,5 +120,5 @@ expectParseFailure s = void . scope (Text.unpack s) . expectLeft . P.parse UriPa
 path :: [Text] -> Path
 path = Path . Seq.fromList . fmap NameSegment
 
-sbh :: Text -> Maybe ShortBranchHash
-sbh = Just . ShortBranchHash
+sbh :: Text -> Maybe ShortCausalHash
```
Comment (mitchellwrosen) [inline unison-cli/tests/Unison/Test/UriParser.hs]: sbh

File: unison-share-api/src/Unison/Server/Backend.hs [hunk 17]
```diff
@@ -46,10 +46,10 @@ import qualified Unison.Codebase.Editor.DisplayObject as DisplayObject
 import Unison.Codebase.Path (Path)
 import qualified Unison.Codebase.Path as Path
 import qualified Unison.Codebase.Runtime as Rt
-import Unison.Codebase.ShortBranchHash
-  ( ShortBranchHash,
+import Unison.Codebase.ShortCausalHash
+  ( ShortCausalHash,
   )
-import qualified Unison.Codebase.ShortBranchHash as SBH
+import qualified Unison.Codebase.ShortCausalHash as SBH
```
Comment (mitchellwrosen) [inline unison-share-api/src/Unison/Server/Backend.hs]: SBH

File: unison-share-api/src/Unison/Server/Backend.hs [hunk 18]
```diff
@@ -1142,21 +1142,21 @@ resolveCausalHashV2 codebase h = case h of
   Just ch -> lift $ Codebase.getShallowCausalForHash codebase ch
 
 resolveRootBranchHash ::
-  Monad m => Maybe ShortBranchHash -> Codebase m v a -> Backend m (Branch m)
+  Monad m => Maybe ShortCausalHash -> Codebase m v a -> Backend m (Branch m)
 resolveRootBranchHash mayRoot codebase = case mayRoot of
   Nothing ->
     lift (Codebase.getRootBranch codebase)
   Just sbh -> do
```
Comment (mitchellwrosen) [inline unison-share-api/src/Unison/Server/Backend.hs]: sbh

File: unison-share-api/src/Unison/Server/Backend.hs [hunk 19]
```diff
@@ -1142,21 +1142,21 @@ resolveCausalHashV2 codebase h = case h of
...
+    h <- expandShortCausalHash codebase sbh
     resolveCausalHash (Just h) codebase
 
 resolveRootBranchHashV2 ::
-  Monad m => Codebase m v a -> Maybe ShortBranchHash -> Backend m (V2Branch.CausalBranch m)
+  Monad m => Codebase m v a -> Maybe ShortCausalHash -> Backend m (V2Branch.CausalBranch m)
 resolveRootBranchHashV2 codebase mayRoot = case mayRoot of
   Nothing ->
     lift (Codebase.getShallowRootCausal codebase)
   Just sbh -> do
```
Comment (mitchellwrosen) [inline unison-share-api/src/Unison/Server/Backend.hs]: sbh

File: unison-share-api/src/Unison/Server/Endpoints/FuzzyFind.hs [hunk 20]
```diff
@@ -31,7 +31,7 @@ import Unison.Codebase (Codebase)
 import qualified Unison.Codebase as Codebase
 import Unison.Codebase.Editor.DisplayObject
 import qualified Unison.Codebase.Path as Path
-import qualified Unison.Codebase.ShortBranchHash as SBH
+import qualified Unison.Codebase.ShortCausalHash as SBH
```
Comment (mitchellwrosen) [inline unison-share-api/src/Unison/Server/Endpoints/FuzzyFind.hs]: SBH

File: unison-share-api/src/Unison/Server/Endpoints/NamespaceListing.hs [hunk 21]
```diff
@@ -164,13 +164,13 @@ backendListEntryToNamespaceObject ppe typeWidth = \case
 
 serve ::
   Codebase IO Symbol Ann ->
-  Maybe ShortBranchHash ->
+  Maybe ShortCausalHash ->
   Maybe Path.Path ->
   Maybe Path.Path ->
   Backend.Backend IO NamespaceListing
 serve codebase maySBH mayRelativeTo mayNamespaceName = do
```
Comment (mitchellwrosen) [inline unison-share-api/src/Unison/Server/Endpoints/NamespaceListing.hs]: SBH

File: unison-share-api/src/Unison/Server/Endpoints/Projects.hs [hunk 22]
```diff
@@ -136,7 +136,7 @@ serve codebase mayRoot mayOwner = projects
       shallowRootBranch <- case mayRoot of
         Nothing -> lift (Codebase.getShallowRootBranch codebase)
         Just sbh -> do
```
Comment (mitchellwrosen) [inline unison-share-api/src/Unison/Server/Endpoints/Projects.hs]: sbh

File: unison-share-api/src/Unison/Server/Errors.hs [hunk 23]
```diff
@@ -12,7 +12,7 @@ import qualified Data.Text.Lazy.Encoding as Text
 import Servant (ServerError (..), err400, err404, err409, err500)
 import qualified Unison.Codebase.Causal as Causal
 import qualified Unison.Codebase.Path as Path
-import qualified Unison.Codebase.ShortBranchHash as SBH
+import qualified Unison.Codebase.ShortCausalHash as SBH
```
Comment (mitchellwrosen) [inline unison-share-api/src/Unison/Server/Errors.hs]: SBH

File: unison-share-api/src/Unison/Server/Orphans.hs [hunk 24]
```diff
@@ -21,10 +21,10 @@ import qualified U.Util.Hash as Hash
 import Unison.Codebase.Editor.DisplayObject
 import qualified Unison.Codebase.Path as Path
 import qualified Unison.Codebase.Path.Parse as Path
-import Unison.Codebase.ShortBranchHash
-  ( ShortBranchHash (..),
+import Unison.Codebase.ShortCausalHash
+  ( ShortCausalHash (..),
   )
-import qualified Unison.Codebase.ShortBranchHash as SBH
+import qualified Unison.Codebase.ShortCausalHash as SBH
```
Comment (mitchellwrosen) [inline unison-share-api/src/Unison/Server/Orphans.hs]: SBH

**Review by mitchellwrosen — CHANGES_REQUESTED** [review]
Nice cleanup -- LGTM, but there are lots of references to the old acronym in variable names and imports. I think we should clean those up too, then merge asap 🤞

**Review by mitchellwrosen — APPROVED** [review]

**Review by aryairani — APPROVED** [review]
We should rename `ShortHash` to `ShortReferent` at some point too.

**Pull request merged**